File: check-imports.py
import os, re, json

# Common MUI components and icons that are often missed (add more as needed)
MUI_COMPONENTS = {
    'Box','Card','CardContent','Typography','Button','TextField','Grid','Avatar','Chip','IconButton',
    'Dialog','DialogTitle','DialogContent','DialogActions','DialogContentText','FormControl','InputLabel',
    'Select','MenuItem','Skeleton','Alert','Divider','Paper','InputAdornment','List','ListItem','ListItemText',
    'ListItemButton','ListItemIcon','LinearProgress','CircularProgress','Badge','Menu','Drawer','AppBar',
    'Toolbar','Tabs','Tab','Tooltip','Fab','Switch','Checkbox','Radio','RadioGroup','FormControlLabel',
    'FormGroup','Autocomplete','Pagination','Table','TableBody','TableCell','TableContainer','TableHead',
    'TableRow','Stepper','Step','StepLabel','StepContent','Accordion','AccordionSummary','AccordionDetails',
    'Rating','Slider','Container','Link','OutlinedInput','FormHelperText','Collapse','SpeedDial','SpeedDialAction',
    'SpeedDialIcon','ToggleButton','ToggleButtonGroup','Stack','Breadcrumbs','Snackbar','Backdrop','Modal',
    'Fade','Grow','Zoom','Slide','ClickAwayListener','Popover','Hidden'
}

# Read all icons from @mui/icons-material? Too many. We'll detect capitalized tags that are not imported and not common components/HTML.
# Instead, we identify all JSX tags (capitalized), subtract imported names, local components, and HTML tags.
# Then report any remaining capitalized tag that is used.

def find_issues(root):
    issues = []
    for dirpath, _, files in os.walk(root):
        if 'node_modules' in dirpath:
            continue
        for f in files:
            if not f.endswith(('.js','.jsx','.ts','.tsx')):
                continue
            path = os.path.join(dirpath, f)
            try:
                text = open(path, 'r', encoding='utf-8').read()
            except Exception:
                continue
            # Find default/named imports
            imported = set()
            # import { ... } from '@mui/material' or '@mui/icons-material' etc.
            for m in re.finditer(r"import\s+\{([^}]+)\}\s+from\s+['\"]([^'\"]+)['\"]", text):
                names = [n.strip().split(' as ')[0].strip() for n in m.group(1).split(',')]
                imported.update(names)
            # import Something from '...'
            for m in re.finditer(r"import\s+(\w+)\s+from\s+['\"][^'\"]+['\"]", text):
                imported.add(m.group(1))
            # import * as Something
            for m in re.finditer(r"import\s+\*\s+as\s+(\w+)\s+from", text):
                imported.add(m.group(1))

            # Find JSX tags
            tags = set(re.findall(r"<([A-Z][a-zA-Z0-9_]*)", text))
            # Also self-closing / components with dot notation like <X.Y> not needed here
            # Remove imported
            undefined = tags - imported
            # Remove common HTML-like React components (Suspense, Fragment, etc)
            undefined -= {'React','Suspense','Fragment','StrictMode','Profiler','Component','Navigate','Route','Routes','Router','BrowserRouter','Link','NavLink','Outlet','ThemeProvider','CssBaseline','AuthProvider','AuthContext','ErrorBoundary','useNavigate','useLocation','useTheme','useMediaQuery','useAuth','useState','useEffect','useCallback','useMemo','useRef','useContext','useReducer','useLayoutEffect','useId'}
            # Remove local function components? Hard. Instead remove lowercase HTML-like and known hooks.
            # If undefined still has common MUI components/icons, report.
            report = []
            for u in undefined:
                # Likely MUI if in MUI_COMPONENTS
                if u in MUI_COMPONENTS:
                    report.append(u)
            if report:
                issues.append((path, sorted(report)))
    return issues

if __name__ == '__main__':
    root = os.path.join(os.path.dirname(__file__), 'src')
    issues = find_issues(root)
    if issues:
        print('POSSIBLE MISSING IMPORTS:')
        for path, names in issues:
            print(f"{path}: {', '.join(names)}")
    else:
        print('No obvious missing MUI imports found.')

File: check-imports.js
const fs = require('fs');
const path = require('path');

const HTML_LIKE = new Set([
  'Suspense','Fragment','StrictMode','Profiler','Component','Navigate','Route','Routes','Router',
  'BrowserRouter','Link','NavLink','Outlet','ThemeProvider','CssBaseline','AuthProvider','AuthContext',
  'ErrorBoundary','RouterLink',' LocalizationProvider','DatePicker','AdapterDateFns'
]);

function walk(dir) {
  const issues = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (entry.name === 'node_modules') continue;
      issues.push(...walk(full));
    } else if (/\.(js|jsx|ts|tsx)$/.test(entry.name)) {
      const text = fs.readFileSync(full, 'utf-8');
      const imported = new Set();
      const importBlock = text.match(/import\s+\{([^}]+)\}\s+from\s+['"]([^'"]+)['"]/g) || [];
      for (const line of importBlock) {
        const m = line.match(/import\s+\{([^}]+)\}\s+from\s+['"]([^'"]+)['"]/);
        if (!m) continue;
        const names = m[1].split(',').map(n => {
          const parts = n.trim().split(/\s+as\s+/).map(s => s.trim());
          return parts.length > 1 ? parts[1] : parts[0];
        });
        for (const n of names) imported.add(n);
      }
      const defaultImports = text.match(/import\s+(\w+)\s+from\s+['"][^'"]+['"]/g) || [];
      for (const line of defaultImports) {
        const m = line.match(/import\s+(\w+)\s+from/);
        if (m) imported.add(m[1]);
      }
      const starImports = text.match(/import\s+\*\s+as\s+(\w+)\s+from/g) || [];
      for (const line of starImports) {
        const m = line.match(/import\s+\*\s+as\s+(\w+)\s+from/);
        if (m) imported.add(m[1]);
      }

      // Local component definitions
      const localDefs = new Set();
      const compRe = /(?:const|function)\s+([A-Z][a-zA-Z0-9_]*)\s*(?:=|\()/g;
      let cm;
      while ((cm = compRe.exec(text)) !== null) localDefs.add(cm[1]);

      const tags = new Set();
      let m;
      const tagRe = /<([A-Z][a-zA-Z0-9_]*)/g;
      while ((m = tagRe.exec(text)) !== null) tags.add(m[1]);

      const undefinedTags = [...tags].filter(t => {
        if (imported.has(t)) return false;
        if (localDefs.has(t)) return false;
        if (HTML_LIKE.has(t)) return false;
        if (t.startsWith('use') && t.length > 3 && t[3] >= 'A' && t[3] <= 'Z') return false; // hooks
        return true;
      });
      if (undefinedTags.length) issues.push({ file: full, names: undefinedTags });
    }
  }
  return issues;
}

const issues = walk(path.join(__dirname, 'src'));
if (issues.length) {
  console.log('POTENTIALLY MISSING IMPORTS:');
  for (const { file, names } of issues) {
    console.log(`${file}: ${names.join(', ')}`);
  }
} else {
  console.log('No obvious missing imports found.');
}
